### DISCUSSION TITLE
User Input validation with costum fonction from list

### DISCUSSION THREAD

**Discussion by AmirZair** [microsoft/BotFramework-Composer#5640] [Q&A]
I have an "ask question" with a herocard as prompt for text with buttons generated from an imported json file using a function wrote with language generation

> [Herocard
     title = Country Selection
    subtitle = Signature rules
    text = What is your Country ?
    buttons = ${foreach(GetSignatureData(), item, cardActionTemplate('imBack', item['value']['value'], item.key))}
]

the function looks like this : 

> #GetSignatureData()
${json(fromFile('./dataSignature.json'))}

The result is stored in dialog.country

I want to allow the user to choose only what proposed in the buttons generated from the heroCard (i'm traversing a json file using keys proposed in the herocard). For that i want to use a Validation rules, but i'm not allowed to put costum fucntion other then what proposed in [Adaptive expressions](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-concept-adaptive-expressions?view=azure-bot-service-4.0&tabs=arithmetic)

![image](https://user-images.githubusercontent.com/24612735/106123994-e44af280-615a-11eb-8943-9b44c2e1c49d.png)

When trying to to put a validation rule like this one : 

>${contains((createArray(foreach(GetSignatureData(), item, item.key))), this.value)}

 i have this error message :

> GetSignatureData() it’s not a built-in function or a custom function.

Thanks in advance teams :)

**Comment by a-b-r-o-w-n** [comment]
@boydc2014 Does @AmirZair need to add the custom function to the project settings or does that not work for custom functions defined in lg?

**Reply by AmirZair** [reply]
Thank you @a-b-r-o-w-n  for the answer, still waiting for @boydc2014. otherwise, is there a way to catch the exception and considering it as an input validation ? (if exception then ask user to only choose a country from the list below)

**Reply by boydc2014** [reply]
Sorry for the late response, yes @AmirZair , what you can do is to add those custom functions into settings page to tell composer it's a function that is allowed. 

Specifically, the field is `customFunctions`, and we support wildchar in there, so you can do something like

customFunctions: [
   "functionA",
   "myfunction.*"
]

**Reply by AmirZair** [reply]
Thank you for your response @boydc2014, i did add my function "GetSignatureData" to customFunction list, but Composer doesn't seem to evaluate it.
![image](https://user-images.githubusercontent.com/24612735/106785012-3f924e80-664d-11eb-8a15-0c77ae360196.png)


As reminder here is the validation Rule : 

> =contains((foreach(GetSignatureData(), item, item.key)), this.value)

And the function GetSignatureData() : 

> \#GetSignatureData()
> ${json(fromFile('./dataSignature.json'))}`

The error given : 
![image](https://user-images.githubusercontent.com/24612735/106784559-bed35280-664c-11eb-8689-888d85c317c6.png)

Thanks in advance
